cmd/openfga/openfga_test.go

```diff
@@ -0,0 +1,402 @@
+	return s.cleanup
+}
+
+// newOpenFGATester spins up an openfga container with the default service ports
+// exposed for testing purposes. Before running functional tests it is assumed
+// the openfga/openfga container is already built and available to the docker engine.
+func newOpenFGATester(t *testing.T, presharedKey string) (OpenFGATester, error) {
+
+	dockerClient, err := client.NewClientWithOpts(client.FromEnv)
+	require.NoError(t, err)
+
+	containerCfg := container.Config{
+		Env: []string{},
+		ExposedPorts: nat.PortSet{
+			nat.Port("8080/tcp"): {},
+			nat.Port("8081/tcp"): {},
+			nat.Port("3000/tcp"): {},
+		},
+		Image: "openfga/openfga",
+		Cmd:   []string{"run"},
+	}
+
+	hostCfg := container.HostConfig{
+		AutoRemove:      true,
+		PublishAllPorts: true,
+		PortBindings: nat.PortMap{
+			"8080/tcp": []nat.PortBinding{},
+			"8081/tcp": []nat.PortBinding{},
+			"3000/tcp": []nat.PortBinding{},
+		},
+	}
+
+	ulid, err := id.NewString()
+	require.NoError(t, err)
+
+	name := fmt.Sprintf("openfga-%s", ulid)
+
+	cont, err := dockerClient.ContainerCreate(context.Background(), &containerCfg, &hostCfg, nil, nil, name)
+	require.NoError(t, err, "failed to create openfga docker container")
+
+	stopContainer := func() {
+
+		timeout := 5 * time.Second
+
+		err := dockerClient.ContainerStop(context.Background(), cont.ID, &timeout)
+		if err != nil && !client.IsErrNotFound(err) {
+			t.Fatalf("failed to stop openfga container: %v", err)
+		}
+	}
+
+	err = dockerClient.ContainerStart(context.Background(), cont.ID, types.ContainerStartOptions{})
+	require.NoError(t, err)
+
+	t.Cleanup(stopContainer)
+
+	// spin up a goroutine to survive any test panics or terminations to expire/stop the running container
+	go func() {
+		time.Sleep(2 * time.Minute)
+
+		// swallow the error because by this point we've terminated
+		_ = dockerClient.ContainerStop(context.Background(), cont.ID, nil)
+	}()
+
+	containerJSON, err := dockerClient.ContainerInspect(context.Background(), cont.ID)
+	require.NoError(t, err)
+
+	ports := containerJSON.NetworkSettings.Ports
+
+	m, ok := ports["8080/tcp"]
+	if !ok || len(m) == 0 {
+		t.Fatalf("failed to get HTTP host port mapping from openfga container")
+	}
+	httpPort := m[0].HostPort
+
+	m, ok = ports["8081/tcp"]
+	if !ok || len(m) == 0 {
+		t.Fatalf("failed to get grpc host port mapping from openfga container")
+	}
+	grpcPort := m[0].HostPort
+
+	backoffPolicy := backoff.NewExponentialBackOff()
+	backoffPolicy.MaxElapsedTime = 30 * time.Second
+
+	err = backoff.Retry(
+		func() error {
+			var err error
+
+			timeoutCtx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
+			defer cancel()
+
+			_ = timeoutCtx
+			_ = err
+			// todo: use grpc health check here
+
+			return nil
+		},
+		backoffPolicy,
+	)
+	require.NoError(t, err)
+
+	return &serverHandle{
+		grpcPort: grpcPort,
+		httpPort: httpPort,
+		cleanup:  stopContainer,
+	}, nil
+}
+
+func randomID(t *testing.T) string {
```

Comment:
This probably belongs in our `testutils` package instead of here.

- [ ] todo(jon-whit): move this to the `testutils` package.